client/my-sites/plan-features-2023-grid/hooks/use-highlight-adjacency-matrix.ts

```diff
@@ -14,19 +19,31 @@ interface Props {
 	visiblePlans: PlanProperties[];
 	flowName: string;
 	currentSitePlanSlug?: string;
+	selectedPlan?: string;
 }
 
-const useHighlightIndices = ( { visiblePlans, flowName, currentSitePlanSlug }: Props ) => {
+const useHighlightIndices = ( {
+	visiblePlans,
+	flowName,
+	currentSitePlanSlug,
+	selectedPlan,
+}: Props ) => {
 	return visiblePlans.reduce< number[] >( ( acc, { planName }, index ) => {
 		let isHighlight = false;
 
+		const isCurrentPlan = currentSitePlanSlug === planName;
+		const isSuggestedPlan = !! ( selectedPlan && planLevelsMatch( planName, selectedPlan ) );
```

Comment:
I notice that for the highlight labels, we also check whether the plan is available for purchase ([here](https://github.com/Automattic/wp-calypso/blob/9c49dc51529c46f5da45fa8596b0f16f72b72085/client/my-sites/plan-features-2023-grid/hooks/use-highlight-label.ts#L25)). Do we need the same condition here, or otherwise update the other and remove? I think it's best if we keep the logic is identical across these two hooks (and sorry for that! 😬).

One way it can lead to "breakage" is if you are on a higher-tier plan and highlight a lower tier via the URL param - the borders around the badge don't get updated correctly. It's probably not the intended use, but still:

<img width="500" alt="Screenshot 2023-05-26 at 4 47 36 PM" src="https://github.com/Automattic/wp-calypso/assets/1705499/3ef9a9c6-eeef-4e93-bdfe-16ff99057b42">

Reply:
p.s. If you need to port the logic over, I fiddled with it a bit and something like this might help:

Somewhere in [highlight-adjacency-matrix.ts](https://github.com/Automattic/wp-calypso/blob/9c49dc51529c46f5da45fa8596b0f16f72b72085/client/my-sites/plan-features-2023-grid/hooks/use-highlight-adjacency-matrix.ts):

```
const isAvailableForPurchase = ( planName: string ) =>
	createSelector(
		( state ) => {
			const selectedSiteId = getSelectedSiteId( state );
			return selectedSiteId ? isPlanAvailableForPurchase( state, selectedSiteId, planName ) : false;
		},
		[ getSelectedSiteId ]
	);
```

then [inside the loop](https://github.com/Automattic/wp-calypso/blob/9c49dc51529c46f5da45fa8596b0f16f72b72085/client/my-sites/plan-features-2023-grid/hooks/use-highlight-adjacency-matrix.ts#L31):

```
const isSuggestedPlan = 
    selectedPlan && planLevelsMatch( planName, selectedPlan ) && isAvailableForPurchase;
```

Reply:
I tried it but it doesn't work. Seems that `isAvailableForPurchase` always returns true, so I guess it's cached/memoized when it shouldn't. Ideally we should use the `isPlanAvailableForPurchase` selector within the `reduce` callback (or even better: reuse the `useHighlightLabel` hook) but it's not possible because of the hooks restrictions.

Reply:
Yeah, that's the problem here. We either need to violate the "rules-of-hooks" or work it out differently. 

This line was wrong, as it's not even calling the selector 😬 :

```
const isSuggestedPlan = 
    selectedPlan && planLevelsMatch( planName, selectedPlan ) && isAvailableForPurchase;
```


This works for me:

```
const useIsAvailableForPurchase = ( planName: string ) =>
	useSelector( ( state ) => {
		const selectedSiteId = getSelectedSiteId( state );
		return selectedSiteId ? isPlanAvailableForPurchase( state, selectedSiteId, planName ) : false;
	} );
```

and then:

```
		const isSuggestedPlan =
			!! ( selectedPlan && planLevelsMatch( planName, selectedPlan ) ) &&
			useIsAvailableForPurchase( planName );
```

but it violates the rule `React Hook "useIsAvailableForPurchase" cannot be called inside a callback.`